File: Two-Furthest-Houses-With-Different-Colors.cpp
1class Solution {
2public:
3    int maxDistance(vector<int>& colors) {
4        int N = colors.size();
5        int left = 0, right = N - 1;
6        int maxdist = 0;
7        while(right > left){
8            if(colors[left]!=colors[right]){
9                maxdist = max(maxdist,right - left);
10                left++;
11                right = N - 1;
12            }
13            else right--;
14        }
15        return maxdist;
16    }
17};

File: Two-Furthest-Houses-With-Different-Colors.java
1class Solution {
2    public int maxDistance(int[] colors) {
3        int maxdist = 0, N = colors.length;
4        for(int right = N - 1 ; right >= 0 ; right--){
5            if(colors[0]!=colors[right]){
6                maxdist = right;
7                break;
8            }
9        }
10        for(int left = 0 ; left < N ; left++){
11            if(colors[left]!=colors[N-1]){
12                return Math.max(maxdist,N-left-1);
13            }
14        }
15        return 0;
16    }
17}